jobmaster/cvsJob/cvsJobMaster.go

```diff
@@ -171,44 +221,53 @@ func (jm *JobMaster) OnWorkerDispatched(worker lib.WorkerHandle, result error) e
 }
 
 func (jm *JobMaster) OnWorkerOnline(worker lib.WorkerHandle) error {
-	// todo : add the worker information to the sync files map
-	jm.syncFilesInfoMu.Lock()
-	defer jm.syncFilesInfoMu.Unlock()
-	syncInfo, exist := jm.syncFilesInfo[worker.ID()]
+	id, exist := jm.launchedWorkers.Load(worker.ID())
 	if !exist {
-		log.L().Panic("phantom worker found", zap.Any("id", worker.ID()))
+		log.L().Info("job master recovering and get new worker", zap.Any("id", worker.ID()), zap.Any("master id", jm.ID()))
+		if jm.IsMasterReady() {
+			log.L().Panic("job master has ready and a new worker has been created, brain split occurs!")
+		}
+		statusBytes := worker.Status().ExtBytes
+		status := cvsTask.Status{}
+		err := json.Unmarshal(statusBytes, &status)
+		if err != nil {
+			// bad json
+			return err
+		}
+		id = status.TaskConfig.Idx
 	} else {
-		log.L().Info("worker online ", zap.Any("id", worker.ID()))
+		log.L().Info("worker online ", zap.Any("id", worker.ID()), zap.Any("master id", jm.ID()))
 	}
-	syncInfo.handle = worker
+	jm.syncFilesInfo[id.(int)].handle.Store(unsafe.Pointer(&worker))
+	jm.syncFilesInfo[id.(int)].needCreate.Store(false)
+	jm.launchedWorkers.Store(worker.ID(), id.(int))
 	return nil
 }
 
+func getTaskConfig(jobStatus *Status, id int) *cvsTask.Config {
+	return &cvsTask.Config{
+		SrcHost:  jobStatus.SrcHost,
+		DstHost:  jobStatus.DstHost,
+		DstDir:   jobStatus.DstDir,
+		StartLoc: jobStatus.FileInfos[id].Location,
+		Idx:      id,
+	}
+}
+
 func (jm *JobMaster) OnWorkerOffline(worker lib.WorkerHandle, reason error) error {
-	jm.syncFilesInfoMu.Lock()
-	defer jm.syncFilesInfoMu.Unlock()
-	syncInfo, exist := jm.syncFilesInfo[worker.ID()]
+	val, exist := jm.launchedWorkers.Load(worker.ID())
 	log.L().Info("on worker offline ", zap.Any(" worker :", worker.ID()))
 	if !exist {
 		log.L().Panic("bad worker found", zap.Any("message", worker.ID()))
 	}
-	// Force to set worker handle, in case of worker finishes before OnWorkerOnline
-	// is fired and worker handle is missing
-	syncInfo.handle = worker
-	log.L().Info("worker finished", zap.String("worker-id", worker.ID()), zap.Any("status", worker.Status()), zap.Error(reason))
-	// TODO failover
+	id := val.(int)
+	if derrors.ErrWorkerFinish.Equal(reason) {
```

Comment:
if the worker finish normally , the worker should be remove from the jm.syncFilesInfo

Reply:
addressed